sdk/core/azure-core/src/cryptography/md5.cpp

```diff
@@ -119,53 +90,88 @@ namespace Azure { namespace Core { namespace Cryptography {
     }
   }
 
-  std::vector<uint8_t> Md5Hash::OnFinal(const uint8_t* data, std::size_t length)
+  std::vector<uint8_t> OnFinal(const uint8_t* data, std::size_t length)
   {
     OnAppend(data, length);
-    _detail::Md5HashContext* md5Context = static_cast<_detail::Md5HashContext*>(m_md5Context);
+
     std::vector<uint8_t> hash;
-    hash.resize(md5Context->hashLength);
+    hash.resize(m_hashLength);
     NTSTATUS status = BCryptFinishHash(
-        md5Context->hashHandle,
-        reinterpret_cast<PUCHAR>(&hash[0]),
-        static_cast<ULONG>(hash.size()),
-        0);
+        m_hashHandle, reinterpret_cast<PUCHAR>(&hash[0]), static_cast<ULONG>(hash.size()), 0);
     if (!BCRYPT_SUCCESS(status))
     {
       throw std::runtime_error("BCryptFinishHash failed");
     }
     return hash;
   }
 
+public:
+  Md5BCrypt()
+  {
+    m_buffer.resize(GetMD5AlgorithmProvider().ContextSize);
+    m_hashLength = GetMD5AlgorithmProvider().HashLength;
+
+    NTSTATUS status = BCryptCreateHash(
+        GetMD5AlgorithmProvider().Handle,
+        &m_hashHandle,
+        reinterpret_cast<PUCHAR>(&m_buffer[0]),
+        static_cast<ULONG>(m_buffer.size()),
+        nullptr,
+        0,
+        0);
+    if (!BCRYPT_SUCCESS(status))
+    {
+      throw std::runtime_error("BCryptCreateHash failed");
```

Comment:
Ditto on the status code

Reply:
Given there are multiple places that do this, Add a ThrowIfFailed(status) method.

Reply:
something like `ThrowIfFailed` would require to know what message to display next to the `status` code before even knowing if we are throwing or not.

I have updated the code adding the code